### PR TITLE
feat: gradient numpy frontend

### LINKED ISSUE #23053: gradient

### PR DESCRIPTION
**Pull request by sineha17**
# added gradient function to numpy frontend

## out of 5 test cases, 4 pass successfully. paddle backend function expects a tensor attribute but the numpy gradient function uses an array. i'm unable to convert the array to a tensor element for the paddle test case without affecting the input datatype of the other test cases.
please provide some insight on how to fix this error.

close #23053

## Checklist 

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py [hunk 1]
```diff
@@ -370,6 +370,49 @@ def test_numpy_gcd(
     )
 
 
+# gradient
+@handle_frontend_test(
+    fn_tree="numpy.gradient",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=("float32", "float16", "float64"),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:377]: the available dtypes should be a strategy that gets all valid dtypes, you should then set the supported dtypes for the function using one of the decorators specified here https://unify.ai/docs/ivy/overview/deep_dive/data_types.html#supported-and-unsupported-data-types

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:1]: please add support for `edge_order` argument in testing

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py [hunk 2]
```diff
@@ -370,6 +370,49 @@ def test_numpy_gcd(
...
+        max_num_dims=3,
+        min_dim_size=2,
+        max_dim_size=4,
+        valid_axis=True,
+        force_int_axis=True,
+    ),
+    varargs=helpers.ints(
+        min_value=-3,
+        max_value=3,
+    ),
```
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:388]: the `varargs` argument can be one of the following as per the function documentation:
1. single scalar to specify a sample distance for all dimensions.
2. N scalars to specify a constant sample distance for each dimension.
       i.e. `dx`, `dy`, `dz`, ...
3. N arrays to specify the coordinates of the values along each
       dimension of F. The length of the array must match the size of
       the corresponding dimension
4. Any combination of N scalars/arrays with the meaning of 2. and 3.

so please extend the testing strategy to accommodate these cases

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py (hunk 2 above)
Comment (sineha17) [reply ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:388]: okay, i'm working on it. thank you

**Comment by sineha17** [comment]
On adding edge_order argument in testing, an exception is raised stating edge_order should not be greater than axis.shape. For example, if axis=0 and edge_order=1 (default value), there is no error produced but in case of axis=0 and edge_order=2, the said exception is raised. For multidimentional arrays the shape of axis will be greater than 1 and in this case edge_order=2 does not produce any error.
I do not know how to restrict the value of edge_order to be less than axis size in the frontend. So far, I've noticed that not adding the edge_order argument in testing and passing only the default value (edge_order=1) is not throwing any error.

**Comment by sineha17** [comment]
in case of paddle backend, the following error occurs. 

ivy.utils.exceptions.IvyValueError: paddle: gradient: paddle: prod: (InvalidArgument) prod(): argument 'x' (position 0) must be Tensor, but got Array (at /paddle/paddle/fluid/pybind/eager_utils.cc:1005)

I have tried adding function decorators which convert numpy array to tensor, also tried importing ivy.paddle/ivy.torch in test file and calling the funtion paddle.tensor/torch.to_tensor inside an if block when backend=='paddle' and still could not resolve this error.

**Comment by MahmoudAshraf97** [comment]
> On adding edge_order argument in testing, an exception is raised stating edge_order should not be greater than axis.shape. For example, if axis=0 and edge_order=1 (default value), there is no error produced but in case of axis=0 and edge_order=2, the said exception is raised. For multidimentional arrays the shape of axis will be greater than 1 and in this case edge_order=2 does not produce any error. I do not know how to restrict the value of edge_order to be less than axis size in the frontend. So far, I've noticed that not adding the edge_order argument in testing and passing only the default value (edge_order=1) is not throwing any error.

to fix this, you might need to write a more complex testing strategy that samples `edge_order` based on the shape and the `axis` argument.



> in case of paddle backend, the following error occurs.
> 
> ivy.utils.exceptions.IvyValueError: paddle: gradient: paddle: prod: (InvalidArgument) prod(): argument 'x' (position 0) must be Tensor, but got Array (at /paddle/paddle/fluid/pybind/eager_utils.cc:1005)
> 
> I have tried adding function decorators which convert numpy array to tensor, also tried importing ivy.paddle/ivy.torch in test file and calling the funtion paddle.tensor/torch.to_tensor inside an if block when backend=='paddle' and still could not resolve this error.

can you please share the entire error trace?

**Comment by sineha17** [comment]
During handling of the above exception, another exception occurred:
ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:375: in test_numpy_gradient
    fn_tree="numpy.gradient",
ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_miscellaneous.py:404: in test_numpy_gradient
    helpers.test_frontend_function(
ivy_tests/test_ivy/helpers/function_testing.py:747: in test_frontend_function
    ret = get_frontend_ret(
ivy_tests/test_ivy/helpers/function_testing.py:2409: in get_frontend_ret
    with ivy_backend.PreciseMode(precision_mode):
ivy/functional/ivy/general.py:86: in __exit__
    raise exc_val
ivy_tests/test_ivy/helpers/function_testing.py:2410: in get_frontend_ret
    ret = frontend_fn(*args, **kwargs)
ivy/functional/frontends/numpy/func_wrapper.py:502: in _outputs_to_frontend_arrays
    ret = fn(*args, **kwargs)
ivy/functional/frontends/numpy/func_wrapper.py:454: in _inputs_to_ivy_arrays_np
    return fn(*ivy_args, **ivy_kwargs)
ivy/functional/frontends/numpy/mathematical_functions/miscellaneous.py:340: in gradient
    return ivy.gradient(f, spacing=varargs, axis=axis, edge_order=edge_order)
ivy/utils/exceptions.py:362: in _handle_exceptions
    _handle_exceptions_helper(e, type(e))
ivy/utils/exceptions.py:371: in _handle_exceptions_helper
    raise cls(fn.__name__, str(e), include_backend=True)
E   ivy.utils.exceptions.IvyValueError: paddle: gradient: paddle: prod: (InvalidArgument) prod(): argument 'x' (position 0) must be Tensor, but got Array (at /paddle/paddle/fluid/pybind/eager_utils.cc:1005)
E   
E   Falsifying example: test_numpy_gradient(
E       frontend='numpy',
E       backend_fw='paddle',
E       on_device='cpu',
E       dtype_input_axis=(['float64'], [array([-1., -1.])], 0),
E       varargs=0,
E       fn_tree='ivy.functional.frontends.numpy.gradient',
E       test_flags=FrontendFunctionTestFlags(
E           num_positional_args=0,
E           with_out=False,
E           inplace=False,
E           as_variable=[False],
E           native_arrays=[False],
E           test_compile=False,
E           generate_frontend_arrays=False,
E           transpile=False,
E           precision_mode=False,
E       ),
E   )

**Comment by MahmoudAshraf97** [comment]
The problem is solved now, please sync your branch with main and try again

**Comment by sineha17** [comment]
> The problem is solved now, please sync your branch with main and try again

Will do, thank you for the update.

**Comment by MahmoudAshraf97** [comment]
Hi, are you still working on this?

**Comment by sineha17** [comment]
> Hi, are you still working on this?

in my local, all the test cases are passing. I've committed all the changes here.

**Review by Ishticode — APPROVED** [review]
@sineha17 
Thanks for the PR and following through with the instruction above. I've tested locally and tests are passing. Apologies for the delays in merging on our end. Thank you very much.

**Pull request merged**